### ISSUE TITLE
Events API showing messages with odd order

### ISSUE THREAD

**Issue by mcdee** [Consensys-Incorporated/teku#7361]
### Description
`/events` event stream is providing `head` messages before `block` messages.

### Steps to Reproduce (Bug)

```
curl -H 'Accept: text/event-stream' -s 'http://localhost:5052/eth/v1/events?topics=head,block'
```

**Expected behavior:**

The `block` event should be emitted when the block is received by the node, and the `head` event after the block has been validated and processed.  As such, the `block` event should come first and the `head` event later.

**Actual behavior:**

The `head` event precedes the `block` event, for example:

```
curl -H 'Accept: text/event-stream' -s 'http://localhost:5052/eth/v1/events?topics=head,block'
: ready
event: head
data: {"slot":"6071990","block":"0xa88f49f19299d7b8408a0ede7eb99c6babd340a5cfcf49b69c2c87568237b7bb","state":"0xa87ad7c4ea2adbc0bdec5791e195a0200e1edd9315a01125834f935887d9
7234","epoch_transition":false,"previous_duty_dependent_root":"0x94c252a2299d3896967a2fbe0ac0e3141d1355843107214be3c4c2c64137a51c","current_duty_dependent_root":"0xb2b98204
87473405fb45891321d674d5bcbbd8523afe1a090df66c0b7b82f1b2","execution_optimistic":false}

event: block
data: {"slot":"6071990","block":"0xa88f49f19299d7b8408a0ede7eb99c6babd340a5cfcf49b69c2c87568237b7bb","execution_optimistic":false}
```

**Frequency:**

Happens all the time (tested on goerli nodes).

### Versions

```
teku/v23.6.1/linux-x86_64/-privatebuild-openjdk64bitservervm-java-17
```

**Comment by rolfyone** [comment]
Thanks for raising, we'll take a look.

**Comment by rolfyone** [comment]
The order here is basically just the order of actual processing, because head update happens while we're processing the block, and we notify of a block once we've completely imported and updated fork choice...

I could potentially delay notification that head has updated, but it'd be specifically for this event stream, most of the rest of the system really does want head changes asap, so it'd be a synthetic delay in notifying external parties of head being updated...

Other than that i'd have to assume a block is going to completely import correctly and notify that earlier, but that could also be inaccurate if incorporating the block fails...
I'm not really sold on either answer, so I figured I'd update the ticket and see how strongly you feel about this being an odd order...

**Comment by rolfyone** [comment]
removed the bug label because it's less of a bug if this is how we're processing the data, although we may still change behaviour if we need to...

**Comment by mcdee** [comment]
When you say "head update happens while we're processing the block",  does this mean you are updating state while going through the P2P validation as per https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/p2p-interface.md#beacon_block ?

If not, then it makes sense to generate the block event once this has passed and the block has been received (prior to any processing of the block that impacts state), and then the head event once the block has been processed.

**Comment by rolfyone** [comment]
That makes sense, will persue that avenue...

**Comment by rolfyone** [comment]
one weird thing, if i just validate but not import a block, then the 'optimistic' flag would only really be able to be based on whether my chain head is optimistic, because i haven't imported the block....

**Comment by rolfyone** [comment]
I think this conversation is ongoing on beacon-apis as to what needs to happen ultimately with this, especially if its used for measurement etc.
For now theres some CLI args that are:
`--beacon-events-block-notify-when-validated-enabled` - sends the event straight after validated (false by default)
`--beacon-events-block-notify-when-imported-enabled` - sends the block event on import, which in our case means you'll get it after head updated. (true by default - existing behaviour).

If you want to have both on you can set the validated flag to enabled, and if you wanted just that one, also set the imported-enabled to false...

We can re-visit this if block events get redone in the beacon-api events...

These flags are available in 23.10.0.
